eclipse-theia/theia
# Upgrade to lerna 5.5.4

#### What it does
Fixes #11737


This PR updates dev-dependency `lerna`, from 4.x to latest 5.x. In the process we get rid of several important security vulnerabilities, that we have carried for a long while (that affect the development environment only, not Theia-based products at runtime). 

We also introduce a new `dev-dependency`, [improved-yarn-audit](https://github.com/djfdyuruiry/improved-yarn-audit), that complements `yarn audit` nicely. Its behaviour is more configurable and its output parse-able, which makes it easier to eventually integrate to CI.

Simple usage examples:
        $> yarn run improved-yarn-audit
        $> yarn run improved-yarn-audit --ignore-dev-deps


#### How to test

 - confirm the repo build locally without issues and that the example application seems to work fine
 - confirm that the previous high and severe security vulnerabilities are gone - there is a new one at level high, related to the `axios` dependency of lerna, used by `nx` (they have a couple of related issues on their project, to use a later `axios`, so it will be fixed soon I hope)
 - (optional - I have done the below myself): test publishing a 'next' and 'latest' release using a local registry. Rough instructions:
 **note: be careful if you happen to have the production registry configured on your machine. Consider moving the corresponding `.npmrc` file temporarily to avoid pushing to real registry by mistake**
    - `$> npm add -g verdaccio`
    - `$> verdaccio`  # start verdaccio in another terminal
    -  _in main terminal, set local registry as temporary default_
    - `$> npm config set registry http://localhost:4873/`
    - `$> yarn config set registry http://localhost:4873/`
    - `$> npm adduser --registry http://localhost:4873/`  # _use bogus user/credentials. e.g.: test/test/test@test.io_
    -  _simulate solid release from the PR commit - select "minor" when prompted_
    - `$> git clean -ffdx && yarn && yarn build:examples && yarn test:theia `
    - `$> npx lerna publish --registry http://localhost:4873 --exact --yes --no-push && yarn -s publish:check`
    - _verify that publishing worked without errors from "publish:check"_
    - _locally build a local Theia application, using the newly locally released Theia (e.g. 1.31.0)_
    - _verify that the local app builds and runs fine_
    - _simulate next release - cleanup any checked-out file left before proceeding_
    - `$> git clean -ffdx && yarn && yarn build:examples && yarn test:theia `
    - `$> npx lerna publish --registry http://localhost:4873 preminor --exact --canary --preid next --dist-tag next --no-git-reset --no-git-tag-version --no-push --yes && yarn -s publish:check`
    - _verify that publishing worked without errors from "publish:check"_
    - _locally build a local Theia application, using the newly locally released Theia (e.g. 1.32.0-next.0)_
    - _verify that the local app builds and runs fine_
    - _when all done, unset local registry (else entries related to it will be included in `yarn.lock` as you continue working on something else:_
    - `$> npm config delete registry http://localhost:4873/`
    - `$> yarn config delete registry http://localhost:4873/`



#### Review checklist


#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
